### PR TITLE
Q gelu

### PR DESCRIPTION
**Pull request by stevenabreu7**

### CODE REVIEW

File: s5/qlayers.py [hunk 1]
```diff
@@ -4,6 +4,202 @@
...
+  """
+  if dtype is None:
+    dtype = jnp.result_type(x) # TODO is this concerning???
+  # promote x to at least float32, this avoids half precision computation
+  # but preserves double or complex floating points
+  dtype = jnp.promote_types(dtype, jnp.float32)
+  x = jnp.asarray(x, dtype)
+
+  mean = jnp.mean(x, axes)
+  mean2 = jnp.mean(_abs_sq(x), axes)
```
Comment (stevenabreu7) [inline s5/qlayers.py:65]: `_abs_aq` is not defined

File: s5/qlayers.py [hunk 2]
```diff
@@ -4,6 +4,202 @@
...
+      param_dtype: The dtype of the parameters.
+      epsilon: Normalization epsilon.
+      use_bias: If true, add a bias term to the output.
+      use_scale: If true, scale the output.
+      bias_init: Initialization function for the bias term.
+      scale_init: Initialization function for the scaling function. 
+    Returns:
+      The normalized input.
+    """
+    reduction_axes = _canonicalize_axes(x.ndim, reduction_axes)
```
Comment (stevenabreu7) [inline s5/qlayers.py:106]: `_canonicalize_axes` is not defined

File: s5/qlayers.py [hunk 3]
```diff
@@ -4,6 +4,202 @@
...
+  epsilon: float = 1e-6
+  dtype: Optional[Dtype] = None
+  param_dtype: Dtype = jnp.float32
+  use_bias: bool = False
+  use_scale: bool = True
+  bias_init: Callable[[PRNGKey, Shape, Dtype], Array] = initializers.zeros
+  scale_init: Callable[[PRNGKey, Shape, Dtype], Array] = initializers.ones
+  reduction_axes: Axes = -1
+  feature_axes: Axes = -1
+  scaling_quantization: int = 8 # TODO: is this appropriate as a default?
```
Comment (stevenabreu7) [inline s5/qlayers.py]: 8bit default is good imo

File: s5/qlayers.py [hunk 4]
```diff
@@ -4,6 +4,202 @@
...
+    dtype: Optional dtype specifying the minimal precision. Statistics
+      are always at least float32 for stability (default: dtype of x).
+    axis_name: Optional name for the pmapped axis to compute mean over.
+    axis_index_groups: Optional axis indices.
+
+  Returns:
+    A pair ``(mean, var)``.
+  """
+  if dtype is None:
+    dtype = jnp.result_type(x) # TODO is this concerning???
```
Comment (stevenabreu7) [inline s5/qlayers.py:58]: is this needed? why?

File: s5/qlayers.py [hunk 5]
```diff
@@ -4,6 +4,202 @@
...
+
+from flax.linen.module import Module, compact, merge_param  # pylint: disable=g-multiple-import
+from jax import lax
+from jax.nn import initializers
+import jax.numpy as jnp
+
+PRNGKey = Any
+Array = Any
+Shape = Tuple[int, ...]
+Dtype = Any  # this could be a real type?
```
Comment (stevenabreu7) [inline s5/qlayers.py]: what does this mean?

File: s5/qlayers.py [hunk 6]
```diff
@@ -26,6 +222,9 @@ class QSequenceLayer(nn.Module):
     dropout: float
     d_model: int
     non_ssm_precision: int
+    use_hard_sigmoid: bool = False # TODO think about this...
```
Comment (stevenabreu7) [inline s5/qlayers.py]: I think we can remove the TODO here, and just run some experiments with this?

File: s5/qlayers.py [hunk 7]
```diff
@@ -50,14 +249,22 @@ def setup(self):
...
-            self.norm = nn.LayerNorm()
+            self.norm = nn.QLayerNorm()
 
         self.drop = nn.Dropout(
             self.dropout,
             broadcast_dims=[0],
             deterministic=not self.training,
         )
 
+        self.gate_op = q_had_maybe(prec)
```
Comment (stevenabreu7) [inline s5/qlayers.py]: `prec` is not defined

**Review by stevenabreu7 — COMMENTED** [review]
A few things are still undefined/incomplete that idk how to fix. Also, how about we use a single parameter for what activation function to use? E.g. a string with gelu | qgelu_approx | hard_sigmoid | etc.?

File: s5/qlayers.py (hunk 5 above)
Comment (kmheckel) [reply s5/qlayers.py]: Comment left over from Flax source

File: s5/qlayers.py (hunk 4 above)
Comment (kmheckel) [reply s5/qlayers.py:58]: They did this in the Flax source, I'm not sure if it's needed but the type promotion was worth flagging for further thought. I don't think it's an issue

File: s5/qlayers.py (hunk 7 above)
Comment (kmheckel) [reply s5/qlayers.py]: My bad, should be `gelu_quant`

**Comment by stevenabreu7** [comment]
it works now, at least for sMNIST. pls reopen if anything breaks for you! @Jegp @kmheckel

**Pull request merged**